.github/workflows/main.yml

```diff
@@ -0,0 +1,109 @@
+name: UX
+on: [push]
+
+jobs:
+  test_keychain:
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v1
+    - name: Set Node Version
+      uses: actions/setup-node@v1
+      with:
+        node-version: 12.16.1
+    - name: Get yarn cache directory path
+      id: yarn-cache-dir-path
+      run: echo "::set-output name=dir::$(yarn cache dir)"
+    - uses: actions/cache@v1
+      id: yarn-cache # use this to check for `cache-hit` (`steps.yarn-cache.outputs.cache-hit != 'true'`)
+      with:
+        path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+        key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+        restore-keys: |
+          ${{ runner.os }}-yarn-
+    - name: Install monorepo deps
+      run: yarn
+    - name: Install Keychain dependencies
+      run: yarn
+      working-directory: ./packages/keychain
+    - name: Keychain tests
+      run: yarn lerna run test --scope @blockstack/keychain
+
+  codecheck:
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v1
+    - name: Set Node Version
+      uses: actions/setup-node@v1
+      with:
+        node-version: 12.16.1
+    - name: Get yarn cache directory path
+      id: yarn-cache-dir-path
+      run: echo "::set-output name=dir::$(yarn cache dir)"
+
+    - uses: actions/cache@v1
+      id: yarn-cache # use this to check for `cache-hit` (`steps.yarn-cache.outputs.cache-hit != 'true'`)
+      with:
+        path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+        key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+        restore-keys: |
+          ${{ runner.os }}-yarn-
+    - name: Monorepo Dependencies
+      run: yarn --frozen-lockfile
+    - name: Lerna Bootstrap
+      run: yarn lerna bootstrap
+    - name: Lint
+      run: yarn lint
+    - name: Depcheck
+      run: yarn lerna run depcheck --parallel
+    - name: Typecheck
+      run: yarn lerna run typecheck --parallel
+
+  publish:
+    needs: [test_keychain, codecheck]
+    runs-on: ubuntu-latest
+    if: github.ref == 'refs/heads/master'
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v1
+    - name: Checkout master
+      run: git checkout master
+    - name: Set Node Version
+      uses: actions/setup-node@v1
+      with:
+        node-version: 12.16.1
+    - name: Get yarn cache directory path
+      id: yarn-cache-dir-path
+      run: echo "::set-output name=dir::$(yarn cache dir)"
+    - uses: actions/cache@v1
+      id: yarn-cache # use this to check for `cache-hit` (`steps.yarn-cache.outputs.cache-hit != 'true'`)
+      with:
+        path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+        key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+        restore-keys: |
+          ${{ runner.os }}-yarn-
+    - name: Install monorepo deps
+      run: yarn
+    - name: Git status
+      run: git status
+    - name: Install package deps
+      run: yarn lerna exec --parallel "yarn"
```

Comment:
We don't need the whole `lerna bootstrap` command, which does `yarn link` in all the packages. We don't need linking because we have TS paths setup. This is faster

Reply:
Wanted to share my thoughts on this, and give it a better shot than on my phone at a restaurant 😅. This is written based on the understanding that the suggestion here is to cross import from directory structures with TS paths.

I've worked with monorepos before that contain modules, similar to what we're doing here, and it worked great. We had rules on what can import what: a "pure" commons folder, and more specific modules that can only import and not be imported etc.

There's a difference here, though, that lerna/our monorepos contains npm packages – and doesn't represent a single product. My understanding of lerna is that we can have the benefits of a monorepo (ease of maintaining versions, single PRs etc), but that we still treat the packages as distinct packages — meaning no cross-importing.

npm packages are great cause they expose an explicit set of functionalities. The abstraction of _this package exports X_. This is lost with cross-importing. Put differently, we're exposing implementations details of one library to another. 
Consider we find the folder structure of one project unsatisfactory and want to refactor it. Cross-importing means we'll need to update paths in all the other libraries that depend on that package. Contrast this with if we use `link` to import the packages. Then it's just the module. So long as with the updated folder structure the same contents is exported, nothing needs to change it the dependents.

We'd be relying on advance build-time configurations to make sure that code isn't duplicated. Say `app` and `connect` both import some code from `keychain`, by default that'll be bundled in with their builds. I'm sure webpack can do some clever stuff to omit paths, but we're leaving open scope for this to happen.

Finally, I think it's confusing to have our development-time version of a library different from the consumer-time version. Even if this doesn't present technical difficulties, it's conceptually more difficult to grasp. Within `ux` we could import _anything_ from our `ui` package. But on, say, a marketing website that doesn't make it into the repo but consumes the library, only what's exported is available.

For the sake of a few moments saved in terms of build time, I don't think it's worth breaking the abstraction of npm packages and exposing implementation details. I'd compare this to saying, within an object oriented design, you can't use any private members, everything can access everything. It breaks the rules of encapsulation.

The alternative here is small in terms of code changes, but massive in terms of architecture. We either use the link functionality, or perhaps we can even achieve the same with TS paths, but pointing to the build output folder.
If we ever need something from one library, and it's not exported, we ask ourselves whether it should be exported, and if so, make sure it's properly exported from that package.

**Bonus point**: we'll also have cleaner/shorter import paths as imports are only ever the package name, and don't contain further paths.

---------------------------------------

Sorry for the wall of text guys, turned into an essay. Hope I understood you correctly chatting on Slack, and that this wasn't what you were suggesting anyway.